Introduce mechanism to specify additional locations for the common cl…

kiril-keranov opened the pull request:
## Summary

Add possibility to specify additional locations to be class loaded by the Tomcat common classloader. 
The default location that the Tomcat common classloader uses to classload from is the `$catalina.base/lib` folder. In order to consider additionally specified locations, the `common.loader` property in the default `catalina.properties` file is modified:
`common.loader=${ADD_TO_COMMON_LOADER},... ` in order to be able to append extra locations for classloading by populating `ADD_TO_COMMON_LOADER`.
The modified `catalina.properties` config file is copied along with the other resource configurations on base layer contribution.
As the `catalina.properties` differs a bit in various major Tomcat releases, modified `catalina.properties` resource files are put for every major Tomcat release that [Paketo Apache Tomcat](https://github.com/paketo-buildpacks/apache-tomcat) supports currently.
`CATALINA_OPTS` default options are modified also with `-DADD_TO_COMMON_LOADER=${ADD_TO_COMMON_LOADER}` in order to reflect the additional locations specified with `ADD_TO_COMMON_LOADER` environment in the `catalina.properties` on Tomcat startup.

**Note**: If `ADD_TO_COMMON_LOADER` is not set adding an empty location to `catalina.properties` won't bother the common classloader to class load its default locations
**Note**: If `ADD_TO_COMMON_LOADER` is having an invalid folder/path specified a warning like the following will be issued on Tomcat startup:
```
12:12:03.933 [main] WARN org.apache.catalina.startup.ClassLoaderFactory -- Problem with directory [/workspace/non-existing-folder]
```
but other locations subject of the common loader won't be affected.

## Use Cases

The mechanism enables other buildpack(s) that contribute stuff in their corresponding layers and append to the `ADD_TO_COMMON_LOADER` launch env in order to specify their layer locations to be loaded by the Tomcat common classloader. This can be done by several buildpacks independently in a loosely coupled manner without knowing or appending to one external configuration package.

## Checklist
<!-- Please confirm the following -->
* [x] I have viewed, signed, and submitted the Contributor License Agreement.
* [ ] I have linked issue(s) that this PR should close using keywords or the Github UI (See [docs](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue))
* [ ] I have added an integration test, if necessary.
* [x] I have reviewed the [styleguide](https://github.com/paketo-buildpacks/community/blob/main/STYLEGUIDE.md) for guidance on my code quality.
* [x] I'm happy with the commit history on this PR (I have rebased/squashed as needed).

dmikusa commented:
A few small issues come to mind...

1. The env variable should be prefixed with `BPI_` since this would be targeted for other buildpacks only, not end users.
2. I'd probably want a name like `BPI_TOMCAT_ADDITIONAL_COMMON_JARS` so it matches closely with the other setting that @loewenstein mentioned. But I'm open to other names if there's a strong reason.
3. I don't like pulling in all those config files from Tomcat, when we're only changing one line. It makes it harder for someone to look and understand what the buildpacks are doing specifically. I would rather see the buildpack take the existing file & update the common loader property and append the value it needs to append.
4. I worry about this being used in conjunction with [external configs](https://github.com/paketo-buildpacks/apache-tomcat?tab=readme-ov-file#external-configuration-package). If a user supplies external config that includes `catalina.properties` then this feature would silently fail. I think my previous comment/suggest of having the buildpack update the file would mitigate this concern. 

Last question, could an upstream buildpack use the external configuration feature to do the same thing? I believe that can include and overlay jars into the common loader location. I'm curious if you've looked into this and if there are any pros/cons of one approach over the other that you can share. Thanks!

loewenstein-sap commented:
As I already expressed, I agree that this should be prefixed with `BPI_` to mark it somewhat CNB specific and not an end user API (although we can of course not technically prevent misusing this) and `TOMCAT_` as it is an env var for the tomcat-build pack.
Regarding `COMMON_LOADER` vs. `ADDITIONAL_COMMON_JARS` - I would favor naming that makes sense if you know Tomcat. We could even think about deprecating `BPI_TOMCAT_ADDITIONAL_JARS` in favor of this proposal - so no naming alignment would be needed.
But all in all, I have no strong opinion an the specifics of the name of this env variable.

I agree that adding a full copy of `catalina.properties` with a single line fix might get out of hand pretty quickly. Unless we would want to support something like `BP_CATALINA_PROPS_{key}={value}` or even - looking at the [environment-variables buildpack](https://github.com/paketo-buildpacks/environment-variables) for inspiration support `APPEND`, `DEFAULT`, `OVERRIDE` and `PREPEND`, then we'd need to take the full scope of the file in our responsibility I guess. Question remains if that amount of flexibility and the attached complexity is really worth it.

We could start to think about this proposal as a first step towards implementing use cases for external config in a more fine granular way and even explicitly prohibit and even technically prevent using this in combination.

Regarding the last question. I don't think external config allows a modular assembly of buildpacks, so I wouldn't go down the road of implementing this use case with ext config in mind as a solution.

kiril-keranov commented:
@loewenstein-sap @dmikusa Thanks for the comments and suggestions!
- I changed the var name to `BPI_TOMCAT_ADDITIONAL_COMMON_JARS` currently
- I removed the pulled `catalina.properties` that were prepared for every major Tomcat release and did the following instead:
  - The default `catalina.properties` is available in the home layer where the corresponding Tomcat version archive is downloaded and expanded in. If no `catalina.properties` is available for `catalina.base` this file is copied from home to base layer. The line for the `common.loader` property is then altered to `common.loader=${BPI_TOMCAT_ADDITIONAL_COMMON_JARS},...`
  - If `catalina.properties` is provided through an [external package](https://github.com/paketo-buildpacks/apache-tomcat?tab=readme-ov-file#external-configuration-package), no copying from home to base is executed but the `common.loader` property is altered in the custom provided `catalina.properties` delivered with the [external package mechanism](https://github.com/paketo-buildpacks/apache-tomcat?tab=readme-ov-file#external-configuration-package) and placed in `catalina.base/conf`.
I believe this covers the case of this being used in conjunction with external configs.
We might also prohibit using them in combination as suggested.

modulo11 commented:
Is editing/copying the `catalina.properties` file actually necessary? Wouldn't adding `-Dcommon.loader=${BPI_TOMCAT_ADDITIONAL_COMMON_JARS},<property read from config>` to `CATALINA_OPTS` be enough?

dmikusa commented on internal/util/replaceInCatalinaProps.go:
```diff
@@ -0,0 +1,27 @@
+
+	lines := strings.Split(string(input), "\n")
+
+	for i, line := range lines {
+		if strings.Contains(line, "common.loader=") {
+			lines[i] = strings.Replace(lines[i], "common.loader=", "common.loader=${BPI_TOMCAT_ADDITIONAL_COMMON_JARS},", 1)
+		}
+	}
```
The trouble is that this *could* be a multi-line field. https://stackoverflow.com/a/8978515/1585136

We should collect multi-line fields for input, as a user could supply a custom catalina.properties file. I don't really care how it writes it back out.

dmikusa replied:
Ah ha, ok. I'm fine with that. We should probably double quote it. If that's not happening. I didn't check the code, just your example above.

```
common.loader="${BPI_TOMCAT_ADDITIONAL_COMMON_JARS}","${catalina.base}/lib",\
"${catalina.base}/lib/*.jar",\
"${catalina.home}/lib",\
"${catalina.home}/lib/*.jar"
```

dmikusa commented:
I like the way you have it now. It will overwrite the file, but a user can just look at the contents of the image to see what's ultimately being used. I think there's a nice simplicity to that.

I was wondering about system properties, and I think there's a flag to make it replace them, but I feel like that gets more confusing as you have to combine system properties and the config files to get a working set of the configuration. I'm open to other opinions though, if someone feel strongly about it.

dmikusa commented:
Sorry, looks like Easy CLA is not OK with things. Can you check on that? We have to have that approved before we can merge. Thanks

kiril-keranov commented:
Easy CLA is ok now

kiril-keranov commented:
@dmikusa Seems pending checks for the last commit are stuck.

dmikusa commented:
Sorry, they needed an @paketo-buildpacks/java-maintainers to approve them and I was OOO. They're running now.

loewenstein-sap commented:
I would probably not recommend to add the deprecation into this PR. @dmikusa thoughts? @modulo11 thoughts?
